src/aspire/utils/rotation.py

```diff
@@ -0,0 +1,287 @@
+        regrot = np.zeros_like(rots)
+        for k in range(K):
+            R = rots[k, :, :]
+            if flag == 1:
+                R = J @ R @ J
+            regrot[k, :, :] = Q_mat.T @ R
+        aligned_rots = copy.copy(self)
+        aligned_rots.rot_matrices = regrot
+        return aligned_rots
+
+    @staticmethod
+    def get_rots_mse(rots_reg, rots_ref):
+        """
+        Calculate MSE between the estimated orientations to reference ones.
+
+        :param rots_reg: The estimated Rotation object after alignment
+             with data matrices in the form of a n-by-3-by-3 array.
+        :param rots_ref: The reference Rotation object.
+        :return: The MSE value between two sets of rotations.
+        """
+        rots_reg = rots_reg.rot_matrices
+        rots_ref = rots_ref.rot_matrices
+        ensure(rots_reg.shape == rots_ref.shape,
+               'Two sets of rotations must have same dimensions.')
+        K = rots_reg.shape[0]
+
+        diff = np.zeros(K)
+        mse = 0
+        for k in range(K):
+            diff[k] = norm(rots_reg[k, :, :] - rots_ref[k, :, :], ord='fro')
+            mse += diff[k] ** 2
+        mse = mse / K
+        return mse
+
+    @staticmethod
+    def common_line_from_rots(r1, r2, ell):
+        """
+        Compute the common line induced by rotation matrices r1 and r2.
+
+        :param r1: The first rotation matrix of 3-by-3 array.
+        :param r2: The second rotation matrix of 3-by-3 array.
+        :param ell: The total number of common lines.
+        :return: The common line indices for both first and second rotations.
+        """
+        ut = np.dot(r2, r1.T)
+        alpha_ij = np.arctan2(ut[2, 0], -ut[2, 1]) + np.pi
+        alpha_ji = np.arctan2(ut[0, 2], -ut[1, 2]) + np.pi
+
+        ell_ij = alpha_ij * ell / (2 * np.pi)
+        ell_ji = alpha_ji * ell / (2 * np.pi)
+
+        ell_ij = int(np.mod(np.round(ell_ij), ell))
+        ell_ji = int(np.mod(np.round(ell_ji), ell))
+
+        return ell_ij, ell_ji
+
+    def __getitem__(self, item):
+        return self.data[item]
+
+    def __setitem__(self, key, value):
+        self.data[key] = value
+
+    def __mul__(self, other):
+        if isinstance(other, Rotation):
+            other = other.data
+            output = np.zeros_like(other)
+            for i in range(self.num_rots):
+                output[i] = np.matmul(self.data[i], other[i])
+        else:
+            output = self.data*other
+        return output
+
+    @property
+    def T(self):
+        """
+        Shortcut to get full set of transposed rotation matrices
+        """
+        return self.transpose()
+
+    def transpose(self):
+        """
+        Apply transpose operation to all rotation matrices
+
+        :return: The set of transposed matrices
+        """
+        T = np.zeros((self.num_rots, 3, 3), dtype=self.dtype)
+        for i in range(self.num_rots):
+            T[i] = self.data[i].T
+        return T
+
+    def __str__(self):
```

Comment:
move all the dunder methods towards the top (after __init__). All staticmethods should be clustered together, and either go before regular instance methods, or after (see the rest of the code to see if there's a convention we're following).

Reply:
Done in new commit  9786c56.